### PR TITLE
[fix][DRAFT] fixing a grad norm bug with ShardedDDP and buckets

### PR DESCRIPTION
**Pull request by blefaudeux**
# Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [x] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?

## What does this PR do?
Should fix https://github.com/facebookresearch/fairscale/issues/517, I added a new unit test to make sure that this was fixed, and that this would not be re-introduced, fairly easy to miss. Spotted by @zhengwy888, thank you !

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/nn/data_parallel/test_sharded_ddp_pytorch_parity.py [hunk 1]
```diff
@@ -173,6 +181,13 @@ def closure_sharded(input_tensor=input_tensor):
                 next(ddp_model.parameters()).requires_grad = not next(ddp_model.parameters()).requires_grad
                 check_same_model_params(sharded_ddp_model, ddp_model, f"Rank: {rank} - Trainability refresh {i} broke")
 
+                if clip_grad_norm:
```
Comment (blefaudeux) [inline tests/nn/data_parallel/test_sharded_ddp_pytorch_parity.py:184]: this is extensively tested on the OSS side, but in that case the gradients are not sharded, so this bug was not caught. I added a simple variant here, checking with one norm only but with the sharded gradients

**Comment by blefaudeux** [comment]
Not ready, the unit test is doing a good job it seems.

File: fairscale/nn/data_parallel/sharded_ddp.py [hunk 2]
```diff
@@ -471,7 +477,7 @@ def reduce(*_: Any) -> None:
                                     group=self.process_group,
                                     async_op=True,
                                 ),
-                                callback=None,
+                                callback=cleanup,
```
Comment (zhengwy888) [inline fairscale/nn/data_parallel/sharded_ddp.py:480]: this should also be in _flush_reduce_calls()

Comment (blefaudeux) [reply fairscale/nn/data_parallel/sharded_ddp.py:480]: I forgot about that, I should store the callback in the bucket actually, that would be  cleaner and make it easier to call it from different places. It's not clear to me that it's what the unit test catches though

Comment (blefaudeux) [reply fairscale/nn/data_parallel/sharded_ddp.py:480]: ok, there was something wrong in the unit test, that second part makes sense

**Comment by blefaudeux** [comment]
to give some more context, this showed up because fairseq was not using https://github.com/facebookresearch/fairscale/blob/master/fairscale/optim/oss.py#L270 but its own variant, which does not check for parameter ownership. Not sure how to guard against that in the future, a little too easy to trip over

**Comment by blefaudeux** [comment]
bigger update incoming, couple of new corner cases fixed
